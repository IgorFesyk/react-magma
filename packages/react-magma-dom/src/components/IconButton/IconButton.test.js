import React from 'react';
import { axe } from 'jest-axe';
import { IconButton } from '.';
import { render, fireEvent, cleanup } from 'react-testing-library';

const TEXT = 'Text';

const BASE_ICON_BUTTON_PROPS = {
  handleClick: jest.fn(),
  icon: 'bell'
};

const renderIconTextButton = (myProps = {}) => {
  const props = {
    ...BASE_ICON_BUTTON_PROPS,
    ...myProps
  };

  return render(<IconButton {...props}>{TEXT}</IconButton>);
};

const renderIconNoTextButton = (myProps = {}) => {
  const props = {
    ...BASE_ICON_BUTTON_PROPS,
    ...myProps,
    ariaLabel: TEXT,
    iconOnly: true
  };

  return render(<IconButton {...props} />);
};

describe('Icon Button', () => {
  afterEach(() => {
    cleanup();
  });

  it('should render icon button with default properties', () => {
    const { container } = renderIconNoTextButton();
    const button = container.querySelector('button');

    expect(button).not.toBeNull();
    expect(button).toHaveAttribute('color', 'primary');
    expect(button).toHaveAttribute('shape', 'round');
    expect(button).toHaveStyleRule('border', '0');
  });

  it('should render icon button with passed in properties', () => {
    const color = 'secondary';
    const shape = 'fill';
    const variant = 'outline';
    const size = 'small';
    const { container, debug } = renderIconNoTextButton({
      color,
      shape,
      variant,
      size
    });
    const button = container.querySelector('button');

    expect(button).toHaveAttribute('color', color);
    expect(button).toHaveAttribute('shape', shape);
    expect(button).toHaveStyleRule('border', '2px solid');
    expect(button).toHaveStyleRule('height', '28px');
  });

  it('should render icon button with text', () => {
    const { getByText } = renderIconTextButton({
      textPosition: 'right'
    });

    expect(getByText(TEXT)).not.toBeNull();
  });

  it('should render icon button with text with passed in properties', () => {
    const color = 'secondary';
    const shape = 'fill';
    const variant = 'outline';
    const size = 'small';
    const textTransform = 'none';
    const { container } = renderIconTextButton({
      textPosition: 'left',
      color,
      shape,
      variant,
      size,
      textTransform
    });
    const button = container.querySelector('button');

    expect(button).toHaveAttribute('color', color);
    expect(button).toHaveAttribute('shape', shape);
    expect(button).toHaveStyleRule('border', '2px solid');
    expect(button).toHaveStyleRule('height', '29px');
    expect(button).toHaveStyleRule('text-transform', textTransform);
  });

  it('should disable a button when the passed disabled', () => {
    const { container } = renderIconNoTextButton({ disabled: true });

    expect(container.querySelector('button')).toBeDisabled();
  });

  it('should trigger the passed in function when icon button is clicked', () => {
    const handleClickSpy = jest.fn();
    const { container } = renderIconNoTextButton({
      handleClick: handleClickSpy
    });

    fireEvent(
      container.querySelector('button'),
      new MouseEvent('click', {
        bubbles: true,
        cancelable: true
      })
    );

    expect(handleClickSpy).toHaveBeenCalledTimes(1);
  });

  it('should trigger the passed in function when icon button with text is clicked', () => {
    const handleClickSpy = jest.fn();
    const { container } = renderIconTextButton({
      textPosition: 'left',
      handleClick: handleClickSpy
    });

    fireEvent(
      container.querySelector('button'),
      new MouseEvent('click', {
        bubbles: true,
        cancelable: true
      })
    );

    expect(handleClickSpy).toHaveBeenCalledTimes(1);
  });

  describe('Icon Button classes', () => {
    describe('Shapes', () => {
      it('default icon button', () => {
        const { container } = renderIconNoTextButton();

        expect(container.querySelector('button')).toHaveStyleRule(
          'border-radius',
          '100%'
        );
      });

      it('fill icon button', () => {
        const { container } = renderIconNoTextButton({ shape: 'fill' });

        expect(container.querySelector('button')).toHaveStyleRule(
          'border-radius',
          '5px'
        );
      });

      it('left cap icon button', () => {
        const { container } = renderIconNoTextButton({ shape: 'leftCap' });

        expect(container.querySelector('button')).toHaveStyleRule(
          'border-radius',
          '5px 0 0 5px'
        );
      });

      it('right cap icon button', () => {
        const { container } = renderIconNoTextButton({ shape: 'rightCap' });

        expect(container.querySelector('button')).toHaveStyleRule(
          'border-radius',
          '0 5px 5px 0'
        );
      });
    });

    describe('Sizes', () => {
      it('default icon button', () => {
        const { container } = renderIconNoTextButton();
        const button = container.querySelector('button');

        expect(button).toHaveStyleRule('height', '37px');
        expect(button).toHaveStyleRule('width', '37px');
        expect(container.querySelector('svg')).toHaveAttribute('height', '18');
      });

      it('large icon button', () => {
        const { container } = renderIconNoTextButton({ size: 'large' });
        const button = container.querySelector('button');

        expect(button).toHaveStyleRule('height', '44px');
        expect(button).toHaveStyleRule('width', '44px');
        expect(container.querySelector('svg')).toHaveAttribute('height', '24');
      });

      it('small icon button', () => {
        const { container } = renderIconNoTextButton({ size: 'small' });
        const button = container.querySelector('button');

        expect(button).toHaveStyleRule('height', '28px');
        expect(button).toHaveStyleRule('width', '28px');
        expect(container.querySelector('svg')).toHaveAttribute('height', '14');
      });

      it('default icon button with text left', () => {
        const { container } = renderIconTextButton({
          textPosition: 'left'
        });
        const span = container.querySelector('span');

        expect(span).toHaveStyleRule('padding-right', '10px');
      });

      it('large icon button with text left', () => {
        const { container } = renderIconTextButton({
          size: 'large',
          textPosition: 'left'
        });
        const span = container.querySelector('span');

        expect(span).toHaveStyleRule('padding-right', '15px');
      });

      it('default icon button with text right', () => {
        const { container } = renderIconTextButton({
          textPosition: 'right'
        });
        const span = container.querySelector('span');

        expect(span).toHaveStyleRule('padding-left', '10px');
      });

      it('large icon button with text right', () => {
        const { container } = renderIconTextButton({
          size: 'large',
          textPosition: 'right'
        });
        const span = container.querySelector('span');

        expect(span).toHaveStyleRule('padding-left', '15px');
      });
    });

    it('custom', () => {
      const color = '#cccccc';
      const { container } = renderIconNoTextButton({ style: { color } });
      const button = container.querySelector('button');

      expect(button).toHaveStyle(`color: ${color}`);
    });
  });

  it('Does not violate accessibility standards for icon button', () => {
    const { container, debug } = renderIconNoTextButton();
    return axe(container.innerHTML).then(result => {
      return expect(result).toHaveNoViolations();
    });
  });

  it('Does not violate accessibility standards for icon button with text', () => {
    const { container } = renderIconTextButton();
    return axe(container.innerHTML).then(result => {
      return expect(result).toHaveNoViolations();
    });
  });
});
