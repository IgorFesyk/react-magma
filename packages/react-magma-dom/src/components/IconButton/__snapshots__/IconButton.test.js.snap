// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`IconButton Icon Only Button Snapshot should render with large size 1`] = `
<div>
  <button
    aria-label="Check Icon"
    class="css-jxizg9"
    icon="[object Object]"
    id="testId"
  >
    <svg
      class="icon"
      fill="currentColor"
      height="24"
      viewBox="0 0 29.37 21.88"
      width="24"
      xmlns="http://www.w3.org/2000/svg"
    >
      <path
        d="M26.71,3.06,11.89,17.88,4.4,10.4.87,13.93l11,11L30.24,6.59Z"
        transform="translate(-0.87 -3.06)"
      />
    </svg>
  </button>
</div>
`;

exports[`IconButton Icon Only Button Snapshot should render with small size 1`] = `
<div>
  <button
    aria-label="Check Icon"
    class="css-194jz7x"
    icon="[object Object]"
    id="testId"
  >
    <svg
      class="icon"
      fill="currentColor"
      height="14"
      viewBox="0 0 29.37 21.88"
      width="14"
      xmlns="http://www.w3.org/2000/svg"
    >
      <path
        d="M26.71,3.06,11.89,17.88,4.4,10.4.87,13.93l11,11L30.24,6.59Z"
        transform="translate(-0.87 -3.06)"
      />
    </svg>
  </button>
</div>
`;

exports[`IconButton Icon Only Button Snapshot should render with updated color 1`] = `
<div>
  <button
    aria-label="Check Icon"
    class="css-phsa7g"
    icon="[object Object]"
    id="testId"
  >
    <svg
      class="icon"
      fill="currentColor"
      height="18"
      viewBox="0 0 29.37 21.88"
      width="18"
      xmlns="http://www.w3.org/2000/svg"
    >
      <path
        d="M26.71,3.06,11.89,17.88,4.4,10.4.87,13.93l11,11L30.24,6.59Z"
        transform="translate(-0.87 -3.06)"
      />
    </svg>
  </button>
</div>
`;

exports[`IconButton Icon Only Button Snapshot should render with updated shape 1`] = `
<div>
  <button
    aria-label="Check Icon"
    class="css-e9oou7"
    icon="[object Object]"
    id="testId"
  >
    <svg
      class="icon"
      fill="currentColor"
      height="18"
      viewBox="0 0 29.37 21.88"
      width="18"
      xmlns="http://www.w3.org/2000/svg"
    >
      <path
        d="M26.71,3.06,11.89,17.88,4.4,10.4.87,13.93l11,11L30.24,6.59Z"
        transform="translate(-0.87 -3.06)"
      />
    </svg>
  </button>
</div>
`;

exports[`IconButton Icon Only Button Snapshot should render with updated variant 1`] = `
<div>
  <button
    aria-label="Check Icon"
    class="css-4in4l5"
    icon="[object Object]"
    id="testId"
  >
    <svg
      class="icon"
      fill="currentColor"
      height="18"
      viewBox="0 0 29.37 21.88"
      width="18"
      xmlns="http://www.w3.org/2000/svg"
    >
      <path
        d="M26.71,3.06,11.89,17.88,4.4,10.4.87,13.93l11,11L30.24,6.59Z"
        transform="translate(-0.87 -3.06)"
      />
    </svg>
  </button>
</div>
`;

exports[`IconButton Icon With Text Button Snapshot should render with large size 1`] = `
<div>
  <button
    aria-label="Check Icon"
    class="css-1pgtik7"
    icon="[object Object]"
    id="testId"
  >
    <span
      class="css-1pwkqkn"
    >
      Test Text
       
    </span>
    <svg
      class="icon"
      fill="currentColor"
      height="20"
      viewBox="0 0 29.37 21.88"
      width="20"
      xmlns="http://www.w3.org/2000/svg"
    >
      <path
        d="M26.71,3.06,11.89,17.88,4.4,10.4.87,13.93l11,11L30.24,6.59Z"
        transform="translate(-0.87 -3.06)"
      />
    </svg>
  </button>
</div>
`;

exports[`IconButton Icon With Text Button Snapshot should render with medium size 1`] = `
<div>
  <button
    aria-label="Check Icon"
    class="css-7n2ufd"
    icon="[object Object]"
    id="testId"
  >
    <span
      class="css-1jsubav"
    >
      Test Text
       
    </span>
    <svg
      class="icon"
      fill="currentColor"
      height="16"
      viewBox="0 0 29.37 21.88"
      width="16"
      xmlns="http://www.w3.org/2000/svg"
    >
      <path
        d="M26.71,3.06,11.89,17.88,4.4,10.4.87,13.93l11,11L30.24,6.59Z"
        transform="translate(-0.87 -3.06)"
      />
    </svg>
  </button>
</div>
`;

exports[`IconButton Icon With Text Button Snapshot should render with small size 1`] = `
<div>
  <button
    aria-label="Check Icon"
    class="css-bl58ye"
    icon="[object Object]"
    id="testId"
  >
    <span
      class="css-1jsubav"
    >
      Test Text
       
    </span>
    <svg
      class="icon"
      fill="currentColor"
      height="12"
      viewBox="0 0 29.37 21.88"
      width="12"
      xmlns="http://www.w3.org/2000/svg"
    >
      <path
        d="M26.71,3.06,11.89,17.88,4.4,10.4.87,13.93l11,11L30.24,6.59Z"
        transform="translate(-0.87 -3.06)"
      />
    </svg>
  </button>
</div>
`;

exports[`IconButton Icon With Text Button Snapshot should render with updated color 1`] = `
<div>
  <button
    aria-label="Check Icon"
    class="css-18myfl8"
    icon="[object Object]"
    id="testId"
  >
    <span
      class="css-1jsubav"
    >
      Test Text
       
    </span>
    <svg
      class="icon"
      fill="currentColor"
      height="16"
      viewBox="0 0 29.37 21.88"
      width="16"
      xmlns="http://www.w3.org/2000/svg"
    >
      <path
        d="M26.71,3.06,11.89,17.88,4.4,10.4.87,13.93l11,11L30.24,6.59Z"
        transform="translate(-0.87 -3.06)"
      />
    </svg>
  </button>
</div>
`;

exports[`IconButton Icon With Text Button Snapshot should render with updated shape 1`] = `
<div>
  <button
    aria-label="Check Icon"
    class="css-7n2ufd"
    icon="[object Object]"
    id="testId"
  >
    <span
      class="css-1jsubav"
    >
      Test Text
       
    </span>
    <svg
      class="icon"
      fill="currentColor"
      height="16"
      viewBox="0 0 29.37 21.88"
      width="16"
      xmlns="http://www.w3.org/2000/svg"
    >
      <path
        d="M26.71,3.06,11.89,17.88,4.4,10.4.87,13.93l11,11L30.24,6.59Z"
        transform="translate(-0.87 -3.06)"
      />
    </svg>
  </button>
</div>
`;

exports[`IconButton Icon With Text Button Snapshot should render with updated text transform 1`] = `
<div>
  <button
    aria-label="Check Icon"
    class="css-xf1vbq"
    icon="[object Object]"
    id="testId"
  >
    <span
      class="css-1jsubav"
    >
      Test Text
       
    </span>
    <svg
      class="icon"
      fill="currentColor"
      height="16"
      viewBox="0 0 29.37 21.88"
      width="16"
      xmlns="http://www.w3.org/2000/svg"
    >
      <path
        d="M26.71,3.06,11.89,17.88,4.4,10.4.87,13.93l11,11L30.24,6.59Z"
        transform="translate(-0.87 -3.06)"
      />
    </svg>
  </button>
</div>
`;

exports[`IconButton Icon With Text Button Snapshot should render with updated variant 1`] = `
<div>
  <button
    aria-label="Check Icon"
    class="css-ebgli1"
    icon="[object Object]"
    id="testId"
  >
    <span
      class="css-1jsubav"
    >
      Test Text
       
    </span>
    <svg
      class="icon"
      fill="currentColor"
      height="16"
      viewBox="0 0 29.37 21.88"
      width="16"
      xmlns="http://www.w3.org/2000/svg"
    >
      <path
        d="M26.71,3.06,11.89,17.88,4.4,10.4.87,13.93l11,11L30.24,6.59Z"
        transform="translate(-0.87 -3.06)"
      />
    </svg>
  </button>
</div>
`;
